plan/gen_physical_plans.go

```diff
@@ -140,9 +140,122 @@ func (p *LogicalJoin) getMergeJoin(prop *requiredProp) []PhysicalPlan {
 			joins = append(joins, mergeJoin)
 		}
 	}
+	// If TiDB_SMJ hint is existed && no join keys in children property,
+	// it should to enforce merge join.
+	if len(joins) == 0 && (p.preferJoinType&preferMergeJoin) > 0 {
+		return p.getEnforcedMergeJoin(prop)
+	}
+
 	return joins
 }
 
+func getNewJoinKeysByOffsets(oldJoinKeys []*expression.Column, offsets []int) []*expression.Column {
+	newKeys := make([]*expression.Column, 0, len(oldJoinKeys))
+	for _, offset := range offsets {
+		newKeys = append(newKeys, oldJoinKeys[offset])
+	}
+	for pos, key := range oldJoinKeys {
+		isExist := false
+		for _, p := range offsets {
+			if p == pos {
+				isExist = true
+				break
+			}
+		}
+		if !isExist {
+			newKeys = append(newKeys, key)
+		}
+	}
+	return newKeys
+}
+
+func getNewEqualConditionsByOffsets(old []*expression.ScalarFunction, offsets []int) []*expression.ScalarFunction {
+	new := make([]*expression.ScalarFunction, 0, len(old))
+	for _, offset := range offsets {
+		new = append(new, old[offset])
+	}
+	for pos, condition := range old {
+		isExist := false
+		for _, p := range offsets {
+			if p == pos {
+				isExist = true
+				break
+			}
+		}
+		if !isExist {
+			new = append(new, condition)
+		}
+	}
+	return new
+}
+
+func (p *LogicalJoin) getEnforcedMergeJoin(prop *requiredProp) []PhysicalPlan {
+	// Check whether SMJ can satisfy the required property
+	leftOffsets := make([]int, 0, len(p.LeftJoinKeys))
+	rightOffsets := make([]int, 0, len(p.RightJoinKeys))
+	for _, col := range prop.cols {
+		isExist := false
+		for leftJoinKeyPos, key := range p.LeftJoinKeys {
+			if key.Equal(p.ctx, col) {
+				for i := 0; i < len(leftOffsets)-1; i++ {
+					if leftOffsets[i] == leftJoinKeyPos {
```

Comment:
@zz-jason Consider the following situation. 
```
tidb > create table ta(a bigint, b bigint, c bigint);
Query OK, 0 rows affected (0.12 sec)

tidb > create table tb(a bigint, b bigint, c bigint);
Query OK, 0 rows affected (0.02 sec)

tidb > desc select /*+ TiDB_SMJ(ta, tb) */ * from ta join tb on ta.a=tb.a and ta.b=tb.b and ta.c=tb.c order by ta.c, tb.b, tb.b, ta.a;
+----------------+--------------+-----------------+------+--------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------+----------+
| id             | parents      | children        | task | operator info
                                   | count    |
+----------------+--------------+-----------------+------+--------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------+----------+
| TableScan_18   |              |                 | cop  | table:ta, range:[-inf,+inf], keep order:false
                                   | 10000.00 |
| TableReader_19 | Sort_20      |                 | root | data:TableScan_18
                                   | 10000.00 |
| Sort_20        | MergeJoin_17 | TableReader_19  | root | root.ta.c:asc, root.ta.b:asc, root.ta.a:asc
                                   | 10000.00 |
| Sort_20        | MergeJoin_17 | TableReader_19  | root | root.ta.c:asc, root.ta.b:asc, root.ta.a:asc
                                   | 10000.00 |
| TableScan_21   |              |                 | cop  | table:tb, range:[-inf,+inf], keep order:false
                                   | 10000.00 |
| TableReader_22 | Sort_23      |                 | root | data:TableScan_21
                                   | 10000.00 |
| Sort_23        | MergeJoin_17 | TableReader_22  | root | root.tb.c:asc, root.tb.b:asc, root.tb.a:asc
                                   | 10000.00 |
| MergeJoin_17   |              | Sort_20,Sort_23 | root | inner join, equal:[eq(root.ta.c, root.tb.c) eq(root.ta.b, root.tb.b) eq(root.ta.a, root.tb.a)], left key:root.ta.c, root.ta.b, root.ta.a, right k
ey:root.tb.c, root.tb.b, root.tb.a | 12500.00 |
+----------------+--------------+-----------------+------+--------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------+----------+
7 rows in set (0.00 sec)
tidb > desc select /*+ TiDB_SMJ(ta, tb) */ * from ta join tb on ta.a=tb.a and ta.b=tb.b and ta.c=tb.c order by ta.c, tb.b, tb.c, ta.a;
+----------------+-------------+-----------------+------+---------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------+----------+
| id             | parents     | children        | task | operator info
                                  | count    |
+----------------+-------------+-----------------+------+---------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------+----------+
| TableScan_10   |             |                 | cop  | table:ta, range:[-inf,+inf], keep order:false
                                  | 10000.00 |
| TableReader_11 | Sort_12     |                 | root | data:TableScan_10
                                  | 10000.00 |
| Sort_12        | MergeJoin_9 | TableReader_11  | root | root.ta.a:asc, root.ta.b:asc, root.ta.c:asc
                                  | 10000.00 |
| TableScan_13   |             |                 | cop  | table:tb, range:[-inf,+inf], keep order:false
                                  | 10000.00 |
| TableReader_14 | Sort_15     |                 | root | data:TableScan_13
                                  | 10000.00 |
| Sort_15        | MergeJoin_9 | TableReader_14  | root | root.tb.a:asc, root.tb.b:asc, root.tb.c:asc
                                  | 10000.00 |
| MergeJoin_9    | Sort_6      | Sort_12,Sort_15 | root | inner join, equal:[eq(root.ta.a, root.tb.a) eq(root.ta.b, root.tb.b) eq(root.ta.c, root.tb.c)], left key:root.ta.a, root.ta.b, root.ta.c, right ke
y:root.tb.a, root.tb.b, root.tb.c | 12500.00 |
| Sort_6         |             | MergeJoin_9     | root | root.ta.c:asc, root.tb.b:asc, root.tb.c:asc, root.ta.a:asc
                                  | 12500.00 |
+----------------+-------------+-----------------+------+---------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------+----------+
8 rows in set (0.00 sec)
```

The first query needn't to sort after merge join, but the second one need. Because the `tb.b` is between `ta.c` and `tb.c` in `order by`'s order.

Reply:
I think the second one don't need to sort after merge join, too.  The `on` condition is `ta.a=tb.a and ta.b=tb.b and ta.c=tb.c`, So we have the following equations:

```txt
  order by ta.c, tb.b, tb.c, ta.a;
= order by ta.c, ta.b, ta.c, ta.a;
= order by ta.c, ta.b, ta.a;
```

Reply:
Yeah, you are right. I will fix it.

Reply:
How about adding these test cases into [explain-test](https://github.com/pingcap/tidb/tree/master/cmd/explaintest)?